### DISCUSSION TITLE
BG Value not showing on Xdrip+ main screen

### DISCUSSION THREAD

**Discussion by asadullah1993** [NightscoutFoundation/xDrip#2109] [Q&A]
Hi, I'm unable to get the BG value and white arrow on main screen after calibrating my free style libre sensor on Xdrip+, and for transmitter, I'm using Miaomiao. This happens sometimes, but not always. Any help on how to get the reading back would be highly appreciated. Thanks. 

Below are screenshots for:

> Xdrip+ System Status 

![Screenshot_20220509-163644_xDrip+](https://user-images.githubusercontent.com/55249960/167402113-c4a8271e-d160-4f1c-aafa-bf4fdef06958.png)

> Just before the calibration

![Screenshot_20220509-162931_xDrip+_1](https://user-images.githubusercontent.com/55249960/167402117-587d9ac7-8ed0-4d28-8f20-d80a8312b75b.png)

> After the calibration (no value here, and no white arrow)

![Screenshot_20220509-162942_xDrip+_1](https://user-images.githubusercontent.com/55249960/167402118-e966ea0b-2aa8-47b3-8ce4-ead7765ec31c.png)

**Comment by Navid200** [comment]
@psonnera

**Comment by psonnera** [comment]
Please check calibration data tables https://xdrip.readthedocs.io/en/latest/use/lesscommon/#show-datatables and post the screenshot

**Reply by asadullah1993** [reply]
Sure, Patrick Sonnerat - I recently changed my sensor and it works fine now for some reason. But the last sensor only lasted 3 days due to this issue. I wasn't planning on risking my hypos. Would this still be helpful if I post the screenshot of the data tables? Thanks a lot for the help, though :) !

And you can let me know what you think the reason is for this, even if those tables aren't helpful now? Thanks 👍

**Reply by psonnera** [reply]
I think the reason was an incorrect intercept. Which is exactly tbought for what you feared: not being able to detect lows. Calibrate first when BG is in the lower range, and stable.

**Reply by asadullah1993** [reply]
Hmmmmm. Right - And do we know of a way to fix the intercept after its been entered incorrectly?  Is that the advice you give at the end? Calibrate first when BG is in lower range, and then stabilize. Is that how you fix the intercept problem?

@psonnera , thank you so much for your assistance. Means a lot :) !

**Reply by psonnera** [reply]
Yes. If you can manage to do first calibration at lower range (say 80-120 / 4.5-6.7) it will define the intercept. Then add a higher calibration point and the slope will be defined. To reset the calculation stop sensor: do not stop sensor only reset all calibration and retry.
Some sensors are very difficult to calibrate though.

**Comment by psonnera** [comment]
I would like to come back on this as many users get stuck on it with a missed readings signal when their bridge/sensor is actually connected and xDrip+ is indeed receiving BG but can't display it due to an invalid intercept calibration. No error message, no useful information unless you dig into events log. Many end up uninstalling and reinstalling xDrip+ and say it "fixes it" but it's only because it forced them to recalibrate and it happened at a time BG was lower and the intercept was acceptable,

**Reply by psonnera** [reply]
This behavior is expected as per https://github.com/NightscoutFoundation/xDrip/blob/6171850e1fdec9cfe45a292c61288557557df3cc/app/src/main/java/com/eveningoutpost/dexdrip/Models/Calibration.java#L792 but assumes a plugin is used hence allowing correct display with an invalid calibration. This is (fortunately) not the case for most people.

**Reply by psonnera** [reply]
Example of stuck state
![image](https://user-images.githubusercontent.com/33809361/177703617-968724f5-0ac4-48cd-8de1-6789c6040b6a.png)

**Reply by Navid200** [reply]
I may be misunderstanding what you think we should do.

The logs are clear that something has gone wrong with calibrations.
If we agree on that, I see two possible actions.  
One is to expect user to clear calibrations and attempt again from the logs.
The other is to change the logs or add a dialog to communicate better with the user that something is wrong with calibrations.

Is the second what you think we should do?

Or, are you thinking that we should change the calibration restrictions and make it less restrictive?

Or, are you just documenting this behavior and are not suggesting that we make any change?

**Reply by psonnera** [reply]
I suggest a static toast and checking there is a plugin enabled before accepting an invalid calibration.
In no way safety mechanisms should be changed, I already believe plugins are not safe.
I would like to code it but that's not in my competences.
NB: users don't know how to read logs...

**Reply by Navid200** [reply]
I don't mind coding.

I'm not sure I understand the concept though.
I used G5, in non-native mode, for more than a year and I never had a confusing situation with xDrip's calibration mechanism.  I understood the delay between blood glucose and interstitial fluid glucose impact on calibration and never calibrated when the trend was steep.  I understood that the more calibrations I made, the less weight each calibration had. 

When someone enters a second calibration and claims that xDrip is broken because the second calibration has less impact than the first, it is an example of the user's misunderstanding of those concepts.
If the problem is that, perhaps, you and I need to somehow find a way to improve our documentation of calibration concepts.  
Or, find a way to encourage users to read them.

Is this a problem that mainly happens with Libre?  
If yes, we have to make sure the change only applies to Libre.

No matter what we do, there will always be a way to break the system by doing something wrong.
Is the problem that people do something obviously wrong, in which case, the problem is not xDrip?
Or is the problem that something reasonable that most people do could cause this, in which case, we can improve xDrip?

I'm not disagreeing with you.  I would like to understand the problem before I say I can contribute to fixing it.

**Reply by psonnera** [reply]
I agree with you on calibration prerequisites, the issue when a bad calibration is entered in xDrip+ and lead to an inconsistent intercept, is that no BG values will be displayed anymore and many users don't understand why. It will eventually lead to a missed readings alarm that is untrue as raw data are flowing in but can't get displayed.
This is now more frequent with Libre since G6 rarely use non-native mode (mostly because few can have raw data) but I've seen one G5 user having the same issue, so I'd say it applies globally to non-native algorithm.
Libre sensors were not made to be calibrated and the relationship between raw data and real BG is not always linear. The further you go above the normal range, the more it becomes inaccurate. I agree ideally the solution would be to insist on the fact people should calibrate only when in range but that's not always possible.
Now, I don't think xDrip+ classic calibration algorithm should be modified at all. In fact, I don't think there is anything wrong with xDrip+ calibration method.

My point is: after adding a calibration, or after putting first calibration, no BG data displays in xDrip+ and the app won't give any hint of why if not inside the events log, why not make something visible for the user? A toast message? A static text message?

If you look at it, there is nothing different between bad and good intercept, only a logged message:
```
               } else if (calibration.intercept > CalibrationAbstract.getHighestSaneIntercept()) {
                    UserError.Log.e(TAG, "Got invalid intercept value in xDrip classic algorithm: " + calibration.intercept);
                    calibration.save();
                    newFingerStickData();

                } else {
                    calibration.save();
                    newFingerStickData();
                }
```
So there's no functionality impact.
And if a plugin is used you can event ignore the event...

I'll try to make a dummy PR if you want.

**Reply by Navid200** [reply]
That's a perfect description.  
If raw readings are coming but something is wrong with calibration, we should clearly inform the user.  Thanks for explaining that.  I understand and agree.

A toast message I believe is useless!
Please, for a test, save your settings.  You will see a toast message.  It lasts only a few seconds.
What if the user is not looking at the screen for just those few seconds?

**Reply by PhilTU** [reply]
As a Libre user I find the insistance on "calibration" at the point of startup inconvenient. In trying to help users I see lots of confusion where they are not receiving data points but are presented with information about calibration as that first screen sits waiting to get the data.

I believe long ago you had to take a blood glucose measurement and enter it into a Dexcom to use it, and I had lazily assumed the xDrip+ behaviour was just a hangover from those days. A Libre will give values without such an action and I have often used apps like Glimp where you are instantly up and running and getting data. In some cases they offer "calibration" at the time of entering a BG reading.

So now I think we see users banging in two BG readings that are often very close in time or value resulting in large extrapolation errors that would be avoided by having a minimum interval and perhaps also a minimum delta on the BG reading, but in any case I fell this should all be a) optional and b) not occurring in the fog of war in starting a new sensor / app / device for the first time.

**Reply by psonnera** [reply]
@PhilTU I believe the algorithm was derived from the one used for Dex, yes, this probably happened when LibreAlarm was added to xDrip+ since the calibration model at that time was G5. https://github.com/NightscoutFoundation/xDrip/commit/d7a63e12b77e288cbf119c9f7a4992b21c2a7002#diff-c3b3b634c78160b1b3bf341860ede6fc9b0fd1042d1072a476533e521dcc1ef8 
You'll find few posts related to G4 and G5 calibration in FB and many for Libre and G6 as I believe the main difference is that people were trained to calibrate G4 and G5 whereas Libre people are not. 
Still, calibrating is difficult for many, this has been discussed here https://github.com/NightscoutFoundation/xDrip/discussions/2028
I've been also trying to convince people to perform first calibration in lower normal range for some time but the point is: some never manage to have stable BG when in range.
Your comment on Glimp is interesting but since it's closed source I don't know how it's made. From your description it probably uses and displays raw data until calibration is entered. Glimp's been there before xDrip+ was for Libre sensors and I haven't read of anybody landing in ER due to Glimp... so maybe it's not as crazy as an idea.
But this might require a lot of modifications that I'm not sure will find support in the developers team...
1- allow raw data display at sensor startup in absence of calibration
2- change the algorithm to force a safe slope at first calibration (what is a safe slope? to me it's a slope that won't make LO happen above 40 whereas the current safety system is to avoid making intercept above 40mg/dl) so that a number can be displayed without a safely defined intercept, I mean how low can be a raw reading of 40? Below 40 or not?
3- maintain the rule above unless a calibration is done in the lower range...
Still, not having to calibrate is a preferred option but only available for Libre 2 with OOP2. https://github.com/NightscoutFoundation/xDrip/pull/1887
I myself use phones supporting OOP1 and get the sensor replaced by the vendor when it fails accuracy in agreement with their terms of sales. I only calibrate when needing to get the sensor running for the last few days and delta is more than 20mg/dl.

**Comment by asadullah1993** [comment]
Hi team, do we know how to solve this issue? I've been facing this recently again on a new sensor. I was hoping to get some answers on this, if any. Thanks.
